tests/UnitTests/OptionalFields.hs

```diff
@@ -0,0 +1,122 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+
+module UnitTests.OptionalFields (optionalFields) where
+
+import Control.Applicative (Alternative (..))
+import UnitTests.OptionalFields.Common
+import UnitTests.OptionalFields.Generics (omitGenerics)
+import UnitTests.OptionalFields.TH (omitTH)
+
+optionalFields :: [TestTree]
+optionalFields = [omitGenerics, omitTH, proofOfConcept]
+
+
+-- c.f. https://github.com/haskell/aeson/pull/839#issuecomment-782453060
+data POC = POC
+  { x :: Nullable Int -- Field is required, but can be null.
+  , y :: Undefineable Int -- Field is optional, but cannot be null.
+  , z :: NullOrUndefineable Int -- Field is optional, and can be null.
+  }
```

Comment:
These tests verify that the PR supports your desired behavior (https://github.com/haskell/aeson/pull/839#issuecomment-782453060), @bergmark :)